File: web/src/lib/providers/chatgpt.ts
import OpenAI from "openai";
import { LENGTH_PRESETS, MODEL_MAP, type Length, type Speed } from "../config";
import type { ProviderResponse } from "./types";

const client = new OpenAI();

export async function queryChatGPT(
  prompt: string,
  speed: Speed,
  length: Length,
  { webSearch = true }: { webSearch?: boolean } = {}
): Promise<ProviderResponse> {
  const model = MODEL_MAP.chatgpt[speed];
  const preset = LENGTH_PRESETS[length];

  // When web search is enabled the model needs headroom for internal search
  // tool calls + the final answer.  Floor at 4096 (same approach as Claude).
  const maxOutputTokens = webSearch
    ? Math.max(preset.maxTokens, 4096)
    : preset.maxTokens;

  const t0 = performance.now();
  const response = await client.responses.create({
    model,
    instructions: preset.systemInstruction,
    input: prompt,
    ...(webSearch && { tools: [{ type: "web_search_preview" as const }] }),
    max_output_tokens: maxOutputTokens,
  });
  const latency = (performance.now() - t0) / 1000;

  return {
    engine: "chatgpt",
    model,
    text: response.output_text ?? "",
    latencySeconds: Math.round(latency * 100) / 100,
    inputTokens: response.usage?.input_tokens ?? 0,
    outputTokens: response.usage?.output_tokens ?? 0,
  };
}

File: metallm/providers/chatgpt.py
"""OpenAI / ChatGPT provider adapter."""

from __future__ import annotations

import openai

from metallm.config import LENGTH_PRESETS, MODEL_MAP, Engine, Length, Speed
from metallm.providers.base import BaseProvider, ProviderResponse


class ChatGPTProvider(BaseProvider):
    engine = Engine.CHATGPT

    def __init__(self) -> None:
        # Reads OPENAI_API_KEY from env automatically.
        self._client = openai.AsyncOpenAI()

    async def query(self, prompt: str, speed: Speed, length: Length) -> ProviderResponse:
        model = MODEL_MAP[self.engine][speed]
        preset = LENGTH_PRESETS[length]

        t0 = self._timer()
        response = await self._client.chat.completions.create(
            model=model,
            max_completion_tokens=preset.max_tokens,
            messages=[
                {"role": "system", "content": preset.system_instruction},
                {"role": "user", "content": prompt},
            ],
        )
        latency = self._timer() - t0

        choice = response.choices[0]
        usage = response.usage

        return ProviderResponse(
            engine=self.engine,
            model=model,
            text=choice.message.content or "",
            latency_seconds=round(latency, 2),
            input_tokens=usage.prompt_tokens if usage else 0,
            output_tokens=usage.completion_tokens if usage else 0,
        )
